addons/purchase/models/product.py

```diff
@@ -48,6 +49,14 @@ def action_view_po(self):
         }
         return action
 
+    @api.onchange('company_id')
+    def _onchange_company_id(self):
+        if self.company_id:
+            po_company_ids = self.env['purchase.order.line'].search([('product_id', 'in', self.product_variant_ids.ids), ('company_id', '!=', self.company_id.id)])
+            if po_company_ids:
+                raise ValidationError(_("You may not change the company on this product to %s because it "
+                                        "is already in use by Purchase Orders within this company:\n%s")
+                                        % (self.company_id.display_name, ', '.join(po_company_ids.mapped("order_id").mapped("name"))))
```

Comment:
The list may be long, add `limit=1` to the search above then
```suggestion
                                        "is already in use by the following Purchase Orders within this company:\n%s")
                                        % (self.company_id.display_name, ', '.join(po_company_ids.order_id.name))))
```